### PR TITLE
feat: Allow modifying `docs` url prefix

### LINKED ISSUE #879: Allow other routes than /docs in the url
## 🚀 Feature

Please allow config to set the basic route for docs to something other than `/docs`

## Have you read the [Contributing Guidelines on issues](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes

## Motivation

We do not use Docusaurus only to write docs but guides and tutorials as well.

## Pitch

Provides better customization?

### PR DESCRIPTION
**Pull request by domcorvasce**
siteConfig.js has a new optional field named *docsUrl* which default value is 'docs'.

This change will allow users who uses the library to host guides and tutorials to customize their websites by assign 'docsUrl' values like 'tutorials' or 'guides'.

Fixes #879

## Motivation

See #879 - I think this is a great addition to the project.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yep

## Test Plan

It uses the existing tests to be sure everything's working fine.

## Related PRs

--

### CODE REVIEW

File: lib/core/defaults.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+module.exports = {
+  DOCS_ROUTE: 'docs',
```
[the remark this answers is not in the document]
Comment (domcorvasce) [reply lib/core/defaults.js]: I'll do it ASAP, thanks for the response.

**Comment by domcorvasce** [comment]
I'll do it ASAP

**Comment by domcorvasce** [comment]
Sorry, I'm kinda new to this. Can you please help me trying to figure out what went wrong?

**Comment by markerikson** [comment]
Pretty sure we'd like something like this for Redux / React-Redux .  The existing Redux docs URLs look like https://redux.js.org/recipes/configuringyourstore , with no `/docs/` prefix.  We _could_ set up redirects, but ideally I'd like to have DS generating exactly the same URLs based solely on config before we switch over.

File: v1/website/core/Footer.js [hunk 2]
```diff
@@ -77,28 +77,28 @@ class Footer extends React.Component {
             <h5>Docs</h5>
             <a
               href={`
-                ${this.props.config.baseUrl}docs/${
+                ${this.props.config.baseUrl}${this.props.config.docsUrl}${
```
Comment (endiliey) [inline v1/website/core/Footer.js]: No more `getDocsUrl` first ?

File: v1/website/pages/en/index.js [hunk 3]
```diff
@@ -51,9 +52,9 @@ class HomeSplash extends React.Component {
                   <div className="pluginRowBlock">
                     <Button
                       href={`
-                        ${siteConfig.baseUrl}docs/${
-                        this.props.language
-                      }/installation
+                        ${siteConfig.baseUrl}${getDocsUrl(
```
Comment (endiliey) [inline v1/website/pages/en/index.js]: Can we refactor all of those getDocsUrl(siteConfig.docsUrl) ?

Something like calling the function only once first at above.

```js
const docsUrl = getDocsUrl(siteConfig.docsUrl);
```

Then use it elsewhere

File: v1/website/pages/en/versions.js [hunk 4]
```diff
@@ -36,9 +37,9 @@ function Versions(props) {
                 <th>{latestVersion}</th>
                 <td>
                   <a
-                    href={`${siteConfig.baseUrl}docs/${
-                      props.language
-                    }/installation`}>
+                    href={`${siteConfig.baseUrl}${getDocsUrl(
```
Comment (endiliey) [inline v1/website/pages/en/versions.js]: Same here on refactoring the number of function calls

File: v1/website/versioned_docs/version-1.3.3/api-site-config.md [hunk 5]
```diff
@@ -81,7 +83,9 @@ headerLinks: [
 
 `cname` - The CNAME for your website. It will go into a `CNAME` file when your site is built.
 
-`customDocsPath` - By default, Docusaurus expects your documentation to be in a directory called `docs`. This directory is at the same level as the `website` directory (i.e., not inside the `website` directory). You can specify a custom path to your documentation with this field.
+`customDocsPath` - By default, Docusaurus expects your documentation to be in a directory called `docs`. This directory is at the same level as the
+
+`website` directory (i.e., not inside the `website` directory). You can specify a custom path to your documentation with this field.
```
Comment (endiliey) [inline v1/website/versioned_docs/version-1.3.3/api-site-config.md]: Are we adding `website` on siteConfig ?

File: v2/lib/load/routes.js [hunk 6]
```diff
@@ -1,5 +1,5 @@
 async function genRoutesConfig({docsMetadatas = {}, pagesMetadatas = []}) {
-  function genDocsRoute(metadata) {
+  function gendocsUrl(metadata) {
```
Comment (endiliey) [inline v2/lib/load/routes.js]: Don't edit v2 files. It's unrelated

File: v1/examples/basics/core/Footer.js [hunk 7]
```diff
@@ -10,7 +10,9 @@ const React = require('react');
 class Footer extends React.Component {
   docUrl(doc, language) {
     const baseUrl = this.props.config.baseUrl;
-    return `${baseUrl}docs/${language ? `${language}/` : ''}${doc}`;
+    const docsUrl = this.props.config.docsUrl;
```
Comment (endiliey) [inline v1/examples/basics/core/Footer.js]: Does this work even when `docsUrl` is null ? In which it should have fallbacked to `docs`

File: v1/examples/basics/pages/en/index.js [hunk 8]
```diff
@@ -20,7 +20,9 @@ function imgUrl(img) {
 }
 
 function docUrl(doc, language) {
-  return `${siteConfig.baseUrl}docs/${language ? `${language}/` : ''}${doc}`;
+  return `${siteConfig.baseUrl}${siteConfig.docsUrl}/${
```
Comment (endiliey) [inline v1/examples/basics/pages/en/index.js]: Does this work even when docsUrl is null ? In which it should have fallbacked to docs

**Comment by domcorvasce** [comment]
Thanks for your feedback @endiliey, I'll look at it in-depth ASAP.  I know you must be busy so thanks again for having found out time to reply.

**Comment by domcorvasce** [comment]
Since I'm busy with school and work, I'd like if someone could look at it, too — @wgao19

**Comment by wgao19** [comment]
Hey @domcorvasce yea I'd like to help. Let me pull up a PR to your branch soon.

@endiliey, regarding your comment here:

> In addition, all the user will have to manually edit their url on website/pages/**.js and website/core/Footer.js since any changes we made on Docusaurus won't be ported to userland (unless it's a first install)

We do have hardcoded `docs/` in our doc site navigations now. So even if this PR lands we will have to change in our side in order to implement the customized prefix to pathname. Although I don't see a workaround? I guess the best we can do is to make sure that the changes here falls back to the hardcoded `docs/` so that we don't break existing sites.

Meanwhile, may I ask when will v2 release?

File: v1/website/core/Footer.js (hunk 2 above)
Comment (wgao19) [reply v1/website/core/Footer.js]: Looks like no need cuz `this.props.config` is passed down from `siteConfig` 😅 Although, might need fallback at `server/routing.js`.

File: v1/website/versioned_docs/version-1.3.3/api-site-config.md [hunk 9]
```diff
@@ -97,6 +101,8 @@ customDocsPath: 'website-docs';
 
 `disableTitleTagline` - An option to disable showing the tagline in the title of main pages. Exclude this field to keep page titles as `Title • Tagline`. Set to `true` to make page titles just `Title`.
 
+`docsUrl` - Route of the documentation. By default, /docs is the default route.
```
Comment (endiliey) [inline v1/website/versioned_docs/version-1.3.3/api-site-config.md]: The default value is `docs`

File: v1/website/versioned_docs/version-1.3.3/api-site-config.md (hunk 5 above)
Comment (endiliey) [reply v1/website/versioned_docs/version-1.3.3/api-site-config.md]: I think there is an intended `\n`

File: v1/lib/server/versionFallback.js [hunk 10]
```diff
@@ -178,11 +180,11 @@ function processVersionMetadata(file, version, useVersion, language) {
   const latestVersion = versions[0];
 
   if (!ENABLE_TRANSLATION && !siteConfig.useEnglishUrl) {
-    metadata.permalink = `docs/${
+    metadata.permalink = `${getDocsUrl(siteConfig.docsUrl)}/${
```
Comment (endiliey) [inline v1/lib/server/versionFallback.js]: Let's do

```js
const docsUrl = getDocsUrl(siteConfig.docsUrl);
```

above and then

```js
metadata.permalink = `${docsUrl)}/.....
```

**Comment by markerikson** [comment]
How's this PR looking?  This is the main thing holding us back from using it for the main Redux docs site, because I'd like to have it natively generate URLs that are the same as what we have right now with Gitbook (such as https://redux.js.org/introduction/coreconcepts ).

**Comment by markerikson** [comment]
Custom permalinks would be another viable option, sure.

**Comment by wgao19** [comment]
@endiliey Thanks so much for the work!

**Comment by JoelMarcey** [comment]
We also want to change the documentation updates in this PR to be applied to the master docs instead of version 1.3.3

**Comment by endiliey** [comment]
Alright, I managed to patch this PR and test it out in a fresh project, seems working fine so far. 


One significant change that I made is that now `siteConfig` is being passed properly as `config` props to all user's own React component. Of course it's not just simple props passing, it's being processed before being passed. So if certain fields are not set, it will fill the default value.

https://github.com/facebook/Docusaurus/blob/59b95d759d736ac8957ca5db5fc029159d6cb87e/v1/lib/server/config.js#L19-L28

So now everytime user needs to access `siteConfig`, they don't need to do things like this (in the past)
```js
const siteConfig = require(`${process.cwd()}/siteConfig.js`);
const docsUrl = siteConfig.docsUrl || 'docs';
```

At the end of the day, all user can now access the siteConfig in their React component like this

```js
const React = require('react');

class MyPage extends React.Component {
  render() {
    const {config} = this.props.config;
    return <div>{config.title}</div>;
  }
}

module.exports = MyPage;
```

Hmm anyway,

shipit

Thank you @domcorvasce for starting this and @wgao19 on helping out

**Comment by markerikson** [comment]
Just used this on the React-Redux docs at https://react-redux.js.org .  Thanks!

**Comment by RilDev** [comment]
Good day,

Does anyone know how to rename the `/docs` route on Docusaurus 2? I would like to change this url from `https://my-site.com/docs/topic1/doc1` to `https://my-site.com/something-else/topic1/doc1`.
I looked throught the documentation and googled it to no avail.

Thank you!

**Comment by yangshun** [comment]
Change the value of `routeBasePath` in the docs plugin to `'something-else'`. Here's an example where I used an empty string to remove the base path but you can use any string you desire.

https://github.com/yangshun/tech-interview-handbook/blob/master/website/docusaurus.config.js#L56

**Comment by RilDev** [comment]
@yangshun Thank you very much it works great!

For reference, added a `routeBasePath` in the `presets` and changed `themeConfig`'s `items`'s `to`. Here is the result:

```
module.exports = {
  title: 'Olivétan',
  tagline: 'La première Bible Francaise.',
  url: 'https://olivetan.fr',
  baseUrl: '/',
  onBrokenLinks: 'throw',
  onBrokenMarkdownLinks: 'warn',
  favicon: 'img/favicon.ico',
  organizationName: 'facebook', // Usually your GitHub org/user name.
  projectName: 'docusaurus', // Usually your repo name.
  themeConfig: {
    navbar: {
      title: 'My Site',
      logo: {
        alt: 'My Site Logo',
        src: 'img/logo.svg',
      },
      items: [
        {
          to: 'bible/',
          activeBasePath: 'docs',
          label: 'Docs',
          position: 'left',
        },
        {to: 'blog', label: 'Blog', position: 'left'},
        {
          href: 'https://github.com/facebook/docusaurus',
          label: 'GitHub',
          position: 'right',
        },
      ],
    },
    footer: {
      style: 'dark',
      links: [
        {
          title: 'Docs',
          items: [
            {
              label: 'Style Guide',
              to: 'docs/',
            },
            {
              label: 'Second Doc',
              to: 'docs/doc2/',
            },
          ],
        },
        {
          title: 'Community',
          items: [
            {
              label: 'Stack Overflow',
              href: 'https://stackoverflow.com/questions/tagged/docusaurus',
            },
            {
              label: 'Discord',
              href: 'https://discordapp.com/invite/docusaurus',
            },
            {
              label: 'Twitter',
              href: 'https://twitter.com/docusaurus',
            },
          ],
        },
        {
          title: 'More',
          items: [
            {
              label: 'Blog',
              to: 'blog',
            },
            {
              label: 'GitHub',
              href: 'https://github.com/facebook/docusaurus',
            },
          ],
        },
      ],
      copyright: `Copyright © ${new Date().getFullYear()} My Project, Inc. Built with Docusaurus.`,
    },
  },
  presets: [
    [
      '@docusaurus/preset-classic',
      {
        docs: {
          sidebarPath: require.resolve('./sidebars.js'),
          routeBasePath: 'bible',
          // Please change this to your repo.
          editUrl:
            'https://github.com/facebook/docusaurus/edit/master/website/',
        },
        blog: {
          showReadingTime: true,
          // Please change this to your repo.
          editUrl:
            'https://github.com/facebook/docusaurus/edit/master/website/blog/',
        },
        theme: {
          customCss: require.resolve('./src/css/custom.css'),
        },
      },
    ],
  ],
};
```

**Pull request merged**